design/pool-configuration.md

```diff
@@ -0,0 +1,552 @@
+  NodeSelector      *metav1.LabelSelector
+}
+```
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    nodeSelector:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values: [hostA, hostB]
+```
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    nodeSelectors:
+    - matchExpression:
+      - key: kubernetes.io/hostname
+        operator: In
+        values: [hostA, hostB]
+```
+
+#### Announcing the IP to a subset of BGP neighbors
+
+This concern is now addressed by adding a peers selector to the `BGPAdvertisement` entity.
+By using both the pools and the peers selector, it will be possible to declare the intent of announcing the IPs of a given `AddressPool` only to a subset of BGP peers,
+and only from a subset of nodes.
+
+```go
+type bgpAdvertisement struct {
+    Name              string
+  Pools             []string
+  Peers             []string
+  PoolSelector      map[string]string
+  PeerSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kGPd: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+#### Not announcing the IP
+
+This concern will be addressed by defining an AddressPool not referenced by any Advertisement instance.
+
+#### Using a given AddressPool only for a subset of services
+
+The idea is to be able to reduce the scope of a particular AddressPool to a set of namespaces and services, by adding an optional namespace / service selector.
+
+```go
+type addressPool struct {
+  Name              string
+  Addresses         []string
+  AvoidBuggyIPs     bool
+  AutoAssign        *bool
+  AllocateTo        serviceAllocation
+}
+
+type serviceAllocation struct {
+  Priority          int
+  Namespaces        []string
+  NamespaceSelector map[string]string
+  ServiceSelector   map[string]string
+}
+```
+
+Example CRD
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: AddressPool
+metadata:
+  name: addresspool-sample3
+  namespace: metallb-system
+spec:
+  addresses:
+    - 2002:2:2::1-2002:2:2::100
+    - 172.20.0.100/24
+  autoAssign: false
+  avoidBuggyIPs: true
+  allocateTo:
+    priority: 50
+      namespaces:
+        - namespaceA
+        - namespaceB
+      serviceSelector:
+          tenant: foo
+```
+
+The controller will use the pool with highest priority in case of multiple matches. This is deterministic and will make the behaviour more clear to the users.
```

Comment:
"Higher is better" is counterintuitive to me. "Priority 1" is usually meant as "top priority". Also, lower is usually better in most routing-related matters (interface "metric", administrative distance, protocol-specific "cost" etc.). I'm aware there are exceptions but we probably want to go with the more common case.

Reply:
Fine, but not priority will still mean the lowest possible priority.

Reply:
Yes, I guess that makes sense.

Also, I'm not saying reversing the logic is *necessarily* the way to go, I just shared **my** understanding of this. Would you rather go with "higher is better"?

Reply:
I don't have a strong opinion about that. I guess I was influenced by localpref which I dealt with just recently. As long as we document it, works for me.